cmd/gateway-gcs.go

```diff
@@ -561,266 +570,190 @@ func (l *gcsGateway) DeleteObject(bucket string, object string) error {
+	meta := gcsMultipartMetaName(uploadID)
+	object := l.client.Bucket(bucket).Object(meta)
 
-	info, err := l.PutObject(bucket, multipartKey, size, data, map[string]string{}, sha256sum)
-	return PartInfo{
-		PartNumber:   partID,
-		LastModified: info.ModTime,
-		ETag:         info.ETag,
-		Size:         info.Size,
-	}, err
-}
+	_, err := object.Attrs(l.ctx)
+	if err != nil {
+		return PartInfo{}, gcsToObjectError(traceError(err), bucket, key)
+	}
 
-// ListObjectParts returns all object parts for specified object in specified bucket
-func (l *gcsGateway) ListObjectParts(bucket string, key string, uploadID string, partNumberMarker int, maxParts int) (ListPartsInfo, error) {
-	delimiter := slashSeparator
-	prefix := pathJoin(ZZZZMinioPrefix, fmt.Sprintf("multipart-%s-%s", escape(key), uploadID))
+	var sha256Writer hash.Hash
 
-	it := l.client.Bucket(bucket).Objects(l.ctx, &storage.Query{Delimiter: delimiter, Prefix: prefix, Versions: false})
+	// Generate random ETag.
+	etag := getMD5Hash([]byte(mustGetUUID()))
 
-	isTruncated := false
+	reader := data
 
-	it.PageInfo().Token = toGCSPageToken(toGCSMultipartKey(key, uploadID, partNumberMarker))
-	it.PageInfo().MaxSize = maxParts
+	if sha256sum != "" {
+		sha256Writer = sha256.New()
+		reader = io.TeeReader(data, sha256Writer)
+	}
 
-	nextPartnumberMarker := 0
+	dataName := gcsMultipartDataName(uploadID, etag)
 
-	parts := []PartInfo{}
-	for {
-		if len(parts) >= maxParts {
-			isTruncated = true
-			break
-		}
+	object = l.client.Bucket(bucket).Object(dataName)
 
-		attrs, err := it.Next()
-		if err == iterator.Done {
-			break
-		} else if err != nil {
-			return ListPartsInfo{}, gcsToObjectError(traceError(err), bucket, prefix)
+	w := object.NewWriter(l.ctx)
+	if md5Hex != "" {
+		var err error
+		w.MD5, err = hex.DecodeString(md5Hex)
+		if err != nil {
+			return PartInfo{}, gcsToObjectError(traceError(err), bucket, key)
 		}
+	}
+	_, err = io.Copy(w, reader)
+	if err != nil {
+		return PartInfo{}, gcsToObjectError(traceError(err), bucket, key)
+	}
 
-		if attrs.Prefix != "" {
-			continue
-		}
+	err = w.Close()
+	if err != nil {
+		fmt.Println(err)
+		return PartInfo{}, gcsToObjectError(traceError(err), bucket, key)
+	}
 
-		_, _, partID, err := fromGCSMultipartKey(attrs.Name)
-		if err != nil {
-			continue
-		} else if partID == 0 {
-			// we'll ignore partID 0, it is our zero object, containing
-			// metadata
-			continue
+	if sha256sum != "" {
+		if hex.EncodeToString(sha256Writer.Sum(nil)) != sha256sum {
+			object.Delete(l.ctx)
+			return PartInfo{}, traceError(SHA256Mismatch{})
 		}
+	}
+
+	return PartInfo{
+		PartNumber:   partID,
+		ETag:         etag,
+		LastModified: time.Now().UTC(),
+		Size:         size,
+	}, err
+}
 
-		nextPartnumberMarker = partID
+// ListObjectParts returns all object parts for specified object in specified bucket
+func (l *gcsGateway) ListObjectParts(bucket string, key string, uploadID string, partNumberMarker int, maxParts int) (ListPartsInfo, error) {
```

Comment:
When an upload is ongoing:

```
$ mc cp /tmp/big myminio/pieinthesky
/tmp/big:                   1.06 GB / 2.00 GB ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓█░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░┃  53.12% 2.59 MB/s 6m10s
```

and parts are present:

```
$ mc ls -r myminio/pieinthesky/minio.sys.temp/multipart/
[2017-06-02 18:05:21 PDT]  64MiB 723e9085-370e-4f7f-860c-ce124cbbbf0a/12595c2aa7777506f82f4c8e3b0b41da
[2017-06-02 18:05:22 PDT]  64MiB 723e9085-370e-4f7f-860c-ce124cbbbf0a/2ad03c6771a52c9d1f35ff92a8cd181d
[2017-06-02 18:07:42 PDT]  64MiB 723e9085-370e-4f7f-860c-ce124cbbbf0a/5e16e9beae118e4e15b7cd7d2c76be79
[2017-06-02 18:06:27 PDT]  64MiB 723e9085-370e-4f7f-860c-ce124cbbbf0a/6736dc9075e494ace0c4dcbb5abb188a
[2017-06-02 18:07:39 PDT]  64MiB 723e9085-370e-4f7f-860c-ce124cbbbf0a/98dd172c572cfc7d14190200ed3fc19f
[2017-06-02 18:07:36 PDT]  64MiB 723e9085-370e-4f7f-860c-ce124cbbbf0a/9e07c558e0544d0f85d72fb51ee8d130
[2017-06-02 18:06:34 PDT]  64MiB 723e9085-370e-4f7f-860c-ce124cbbbf0a/b85a2728e7147714fda7adb31a467b46
[2017-06-02 18:05:19 PDT]  64MiB 723e9085-370e-4f7f-860c-ce124cbbbf0a/e65147ee357f042665ca147e81fca981
[2017-06-02 18:04:10 PDT]    53B 723e9085-370e-4f7f-860c-ce124cbbbf0a/gcs.json
```

listing parts of the object is failing, and listing incomplete uploads is failing too:

```
$ mc ls -I myminio/pieinthesky/big
mc: <ERROR> Unable to stat `myminio/pieinthesky/big`. Object does not exist
$ mc ls -I myminio/pieinthesky
$
```